From e028d3201afe2261626216a04a19bca15307e6cc Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 15:21:56 -0400
Subject: [infiniband] ocrdma: Unregister inet notifier when unloading ocrdma

Message-id: <01583662866f23145c3d732d984beb4b0b769d55.1411571481.git.dledford@redhat.com>
Patchwork-id: 96627
O-Subject: [Patch RHEL7 21/48] RDMA/ocrdma: Unregister inet notifier when unloading ocrdma
Bugzilla: 1081067 1088577
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Selvin Xavier <selvin.xavier@emulex.com>

Unregister the inet notifier during ocrdma unload to avoid a panic after
driver unload.

Signed-off-by: Selvin Xavier <selvin.xavier@emulex.com>
Signed-off-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 2d8f57d56f5850a75d6225519e77f75ad9949f25)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_main.c b/drivers/infiniband/hw/ocrdma/ocrdma_main.c
index 8bd9db6..7c504e0 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_main.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_main.c
@@ -539,6 +539,11 @@ static void ocrdma_unregister_inet6addr_notifier(void)
 #endif
 }
 
+static void ocrdma_unregister_inetaddr_notifier(void)
+{
+ unregister_inetaddr_notifier(&ocrdma_inetaddr_notifier);
+}
+
 static int __init ocrdma_init_module(void)
 {
  int status;
@@ -552,13 +557,19 @@ static int __init ocrdma_init_module(void)
 #if IS_ENABLED(CONFIG_IPV6)
  status = register_inet6addr_notifier(&ocrdma_inet6addr_notifier);
  if (status)
-  return status;
+  goto err_notifier6;
 #endif
 
  status = be_roce_register_driver(&ocrdma_drv);
  if (status)
-  ocrdma_unregister_inet6addr_notifier();
+  goto err_be_reg;
 
+ return 0;
+
+err_be_reg:
+ ocrdma_unregister_inet6addr_notifier();
+err_notifier6:
+ ocrdma_unregister_inetaddr_notifier();
  return status;
 }
 
@@ -566,6 +577,7 @@ static void __exit ocrdma_exit_module(void)
 {
  be_roce_unregister_driver(&ocrdma_drv);
  ocrdma_unregister_inet6addr_notifier();
+ ocrdma_unregister_inetaddr_notifier();
  ocrdma_rem_debugfs();
 }
 
-- 
1.7.1